Lib/asyncio/pools.py

```diff
@@ -0,0 +1,166 @@
+"""Support for high-level asynchronous pools in asyncio."""
+
+__all__ = 'ThreadPool',
+
+
+import concurrent.futures
+import functools
+import threading
+import os
+
+from . import events
+from . import exceptions
+from . import futures
+
+
+class AbstractPool:
+    """Abstract base class for asynchronous pools."""
+
+    async def astart(self):
+        raise NotImplementedError
+
+    async def __aenter__(self):
+        await self.astart()
+        return self
+
+    async def aclose(self):
+        raise NotImplementedError
+
+    async def __aexit__(self, exc_type, exc_value, exc_traceback):
+        await self.aclose()
+
+    async def run(self, /, func, *args, **kwargs):
+        """Asynchronously run function *func* using the pool.
+
+        Return a future, representing the eventual result of *func*.
+        """
+        raise NotImplementedError
+
+
+class ThreadPool(AbstractPool):
+    """Asynchronous threadpool for running IO-bound functions.
+
+    Directly calling an IO-bound function within the main thread will block
+    other operations from occurring until it is completed. By using a
+    threadpool, several IO-bound functions can be ran concurrently within
+    their own threads, without blocking other operations.
+
+    The optional argument *concurrency* sets the number of threads within the
+    threadpool. If *concurrency* is `None`, the maximum number of threads will
+    be used; based on the number of CPU cores.
+
+    This threadpool is intended to be used as an asynchronous context manager,
+    using the `async with` syntax, which provides automatic initialization and
+    finalization of resources. For example:
+
+    import asyncio
+
+    def blocking_io():
+        print("start blocking_io")
+        with open('/dev/urandom', 'rb') as f:
+            f.read(100_000)
+        print("blocking_io complete")
+
+    def other_blocking_io():
+        print("start other_blocking_io")
+        with open('/dev/zero', 'rb') as f:
+            f.read(10)
+        print("other_blocking_io complete")
+
+    async def main():
+        async with asyncio.ThreadPool() as pool:
+            await asyncio.gather(
+                pool.run(blocking_io),
+                pool.run(other_blocking_io))
+
+    asyncio.run(main())
+    """
+
+    def __init__(self, concurrency=None):
+        if concurrency is None:
+            concurrency = min(32, (os.cpu_count() or 1) + 4)
```

Comment:
This heuristic looks a bit gratuitous. I think it would be less surprising for the user to have `concurrency = os.cpu_count()` as a default, just like `concurrent.futures` does.

Reply:
@pitrou 

> I think it would be less surprising for the user to have concurrency = os.cpu_count() as a default, just like concurrent.futures does.

ProcessPoolExecutor uses `os.cpu_count()`, but ThreadPoolExecutor uses `min(32, (os.cpu_count() or 1) + 4)` (which is what I based the above on):

https://github.com/python/cpython/blob/02a4d57263a9846de35b0db12763ff9e7326f62c/Lib/concurrent/futures/thread.py#L135-L143

Reply:
Uh, I see. That's the unfortunate result of wanting a default that's "good" for both IO concurrency and parallel execution of CPU-bound...

Reply:
Also to clarify, it looks like that was changed fairly recently (last year) as a part of https://bugs.python.org/issue35279. Previously, ThreadPoolExecutor used `os.cpu_count() * 5`. 

Since @asvetlov and @methane were involved in that change (https://github.com/python/cpython/pull/13618), do either of you have any thoughts regarding the default max threads for asyncio.ThreadPool? Would the current default of ThreadPoolExecutor also make sense here, or should we potentially consider a different value?

Reply:
Discussion in bpo-35279 can be applied to this thireadpool too.  Using same default makes sense.
But why do you calculate the default value here?  Why don't you pass None to ThreadPoolExecutor?

Reply:
> But why do you calculate the default value here? Why don't you pass None to ThreadPoolExecutor?

The reason the maximum is calculated locally is because a primary goal for `asyncio.ThreadPool` is to transition into a native implementation that doesn't rely on `concurrent.futures`, see the discussion in bpo-32309 for more details. So, while we would just pass None to *max_workers* when it's not specified by the user in this specific version, I think it makes more sense to define the maximum locally for the long term.